### ISSUE TITLE
Loading or pretraining may be flawed

### ISSUE THREAD

**Issue by Luca96** [tensorforce/tensorforce#684]
Hi @AlexKuhnle, I'm trying to train a `PPO` agent in chunks just because takes too time (like days) to train it in a single shot. So to speed up training I'm also pretraining it on some (near) optimal experiences.

I ran 128 pretraining episodes and got this nice learning curve (the light blue one):
![baseline_loss_pretraining](https://user-images.githubusercontent.com/11301884/82231353-5234ee00-992d-11ea-8766-cee5baf032d0.png)
The baseline loss drops to ~5000 after 128 episodes. It's still high so I restarted the code expecting to resume where it stopped before... The second run is represented by the _magenta dot_, as you can see the loss is the same as the beginning of the first run, i.e. ~1e+6!

 I regularly save every 32 epochs with the following code:
```python
agent.save(directory=weights_dir, filename='ppo8', format='tensorflow', append=None)
```
 In my `weights_dir` I have the following files:
```
- checkpoint
- ppo8.data-00000-of-00001
- ppo8.index
- ppo8.json
- ppo8.meta
- ppo8.pb
```
The loading code looks like this:
```python
agent = Agent.load(directory=weights_dir, filename='ppo8, format=None, environment=environment)
```
Got any idea? Thanks for help.

PS: I have three more bonus questions :) You can skip them if you don't have time.
1. I noticed that the `empirical-reward` is always some order of magnitude higher than the `episode-reward` although the respective plots have the same shape. Is this normal?

![empirical_vs_episode_reward](https://user-images.githubusercontent.com/11301884/82233689-ad1c1480-9930-11ea-9956-69212ca30057.png)

2. What happens to decaying parameters? Is Tensorforce able to track the current decay step, parameter's value, and so on?
3. How can I understand if training is going well? The baseline (value) loss should be as close as possible to zero. Whilst the policy gradient loss should be positive, negative, or close to zero? And what about the KL-divergences and entropies? 

(sorry for so many questions.)

**Comment by AlexKuhnle** [comment]
I have no good answer for why the curves for continuing pretraining look like this. I hope I can reproduce it with PPO/CartPole. It clearly looks like something is not recovered correctly.

Comment on pretrain: it's a primitive implementation which likely won't work super-well (but should help to bootstrap). There are more sophisticated approaches (and discussion of problems) around inverse RL, learning from demonstrations, etc.

Bonus questions:

1. Episode-reward is the sum of all timestep rewards, whereas empirical-reward is the agent's estimate, so discounted sum of reward horizon plus optional discounted horizon estimate. I also realise that, depending on the agent, these are based on different data (empirical-reward happens during update, episode-reward "on-policy") -- not ideal, will be improved. So it depends on the agent and reward estimation config, could you post the relevant spec values?

2. Yes, there should be a summary "parameters" to track their values.

3. Hard to say... :-)  Losses, entropy, KL-divergence should all decrease generally (losses are minimized, incl policy gradient loss but this one is no "non-negative distance" with well-defined minimum 0, less entropy means more certain decisions, less KL-div means policy convergence). However, these are only tendencies, and even the losses do not behave quite in the way they would for supervised learning (while they're directly minimized, they optimize for a moving target, in a way, so learning progress could even imply increasing losses). I think you're looking at the right things by checking rewards/returns/estimates. If the reward estimation doesn't make sense or seems mis-aligned, then there is presumably little chance the agent will learn well. And the mismatch between episode- and empirical-reward could be worrying, depending on your config. What does entropy and KL look like?

**Comment by Luca96** [comment]
First of all, thank you very much for your soon and exhaustive answers.

1. Here is my `ppo` agent's (pseudo-)spec:
```python
max_episode_timesteps = 400
discount  = 0.99
likelihood_ratio_clipping = linear_decay(initial_value=0.25, steps=10_000, unit='updates')
learning_rate = linear_decay(initial_value=1e-5, steps=10_000, unit='updates')
batch_size = 1
subsampling_fraction = 0.2
optimization_steps = 10
critic_optimizer = # same as policy, with linearly decaying lr = 3e-5
entropy_regularization = 0.0
exploration = linear_decay(initial_value=0.1, steps=200, unit='updates')
variable_noise = 0.0
```
Moreover, the policy and critic networks are the same.

2. Ok, I'll check that!
3.  I only have some plots of the pretraining phase. The best performing agent is the one associated to the light blue curve:
![imgonline-com-ua-twotoone-flvykLA7gFv5](https://user-images.githubusercontent.com/11301884/82334991-fe87da80-99e8-11ea-9c23-50d40940b2ba.jpg)
As you can see, the entropies are steady during the 128 pretraining episodes, and so the KL-divergences are close to 0. The baseline loss drops, while the objective loss reaches a positive value sligtly greater than zero. Seems good, right?

**Comment by AlexKuhnle** [comment]
Two things are a bit weird: First, why is the empirical reward constant? Something related to your environment, by any chance? (Are all episodes the same length, is the overall reward constant?) Second, the losses for the main policy increase, not decrease. However, given that the baseline loss is going down, that probably explains it: The baseline learns a more accurate state value estimate, which in turn affects the loss (particularly, since the overall return contributes multiplicatively, `ratio * Adv`, the difference between estimated value and empirical return `Adv = Q - V` decreasing means the overall loss moves towards zero.

So yes, it doesn't look bad, but we can also see that the main loss is not very straightforward to read for policy gradient models. Is this now better than the problem you mentioned in the beginning?

Also, I assume your `linear_decay` macro implicitly sets the `final_value` to zero?

**Comment by Luca96** [comment]
Well, maybe you're right about the constant empirical reward. These plots refer to pretraining on some fixed-length traces. Also the reward slightly oscillates between 18 and 20, so that's why it looks constant.

 `linear_decay` sets the `final_value` of likelihood_ratio_clipping  and exploration to 0, while for learning rates it's set equal to their initial value divided by 100. So, 1e-7 and 3e-7.

I also tried to save and load the agent with a `saver(...)`, so without doing it explicitly via `agent.save`. I noticed that the agent got saved only at the beginning, and not regularly as specified by `frequency`. That's strange because I remembered that it used to work :/ 
Also the `saver` produces some files that are different from then ones obtained by `agent.save`, despite using the same `"tensorflow"` format. There is a `graph.pbtxt` and `events.out.tfevents.xyz` files in addition.

Anyway, I think it's better to enable saving, and summarising on the `tf2` branch first. So, if I can help somehow with the code (apart from testing it), just let me know what can I do.

**Comment by AlexKuhnle** [comment]
Ah, okay. Regarding saver: Using explicit `agent.save` also creates an act-only Protobuf version of the graph for deployment (`graph.pbtxt`). Maybe not ideal, since you didn't really ask for it to be created, but this will change anyway with the move to SavedModel and Checkpoint in the next version. I have no idea why the `events.out...` file is created, but it doesn't seem to affect anything and is not big (I think), so didn't look into it more. Again, will all change. If `saver` doesn't work correctly anymore, that would be very annoying, and maybe related to recent TF2 changes. One reason more to justify the work for the new version, I guess... :-)

I'll push an update to the branch later. Saving now mostly works well, however, SavedModel saving seems to have some problems I don't yet understand. Summaries also start working, but require more intrusive changes (a few summary options will have to go, at least for now, none of the essential ones though, like the ones above).

One thing comes to mind where help would be great, and which came up in another of your issues: The new Checkpoint feature should make it easily possible to save and load individual modules, so it should now be straightforward to support Keras networks and loading of pre-trained feature extractors, for instance. A few ideas: it would be great, as mentioned in the other issue, to have a `KerasNetwork` class as an alternative to specify networks. It may also be useful to have a `KerasModel/Block` layer which loads a pre-trained "black-box" (based on CheckPoint? And/or even SavedModel?) and wraps it into the layer interface to be used within an RL network architecture.

**Comment by Luca96** [comment]
What kind of issues cause SavedModel? Maybe I can help...

Just, few questions about `KerasNetwork` and `KerasModel`:
1. What should be the role of `KerasNetwork` class? I mean `tf.keras.Model` already have all the necessary machinery to build, load, and save keras models, what else could be necessary?
2. When building networks using only keras API, it's possible to define (sub-)networks first, and then compose them together into one network. This is commonly done to define autoencoders and GANs. 

   In pseudo-code it should look like this:
   ```python
    # build networks first
    encoder = build_encoder()
    decoder = build_decoder()

    # connect the two architectures together
    autoencoder = decoder(encoder.input)

    # now gradients flow from decoder's output to encoder's input.
    ```
    If the input/output dimensions match, one can substitute e.g. the `encoder` with a pre-defined and/or pre-trained architecture available in the `tf.keras.applications` package (e.g. `ResNets`).

    Now, the question is that a `KerasModel` class could offer the same functionalities of a `KerasNetwork` class unless you want to decouple the keras model definition (to `KerasNetwork`) and its effective building or loading (to `KerasModel`).

**Comment by AlexKuhnle** [comment]
If you run the `test_saving` unittest with the next commit, coming soon, you can see it. Some weird:
`tensorflow.python.autograph.impl.api.StagingError: in user code:`
`TensorforceError: Invalid type for NestedDict argument value: <class 
 'tensorflow.python.util.nest._DotString'>.`

1. Maybe not much apart from writing a Tensorforce Network wrapper around it. However, there may be incompatibilities with some of the higher-level API functions, since Keras is built with a supervised setup in mind, and there tend to be details where supervised architectures sometimes clash with Tensorforce's design. Best to find out by trying to build it... :-)

2. With `KerasBlock` and `register`/`retrieve`, this could be possible, although not quite as straightforward as your code. I'm not sure it will be possible to enable it exactly like that, since building the network has to happen at a certain point within the Tensorforce architecture (this is one of the points where supervised and RL/Tensorforce sometimes don't align so well), so depends on how this works with network initialization.

**Comment by Luca96** [comment]
I don't know why but seems that `pygame` doesn't like eager execution :thinking: I got this weird error:
```
Fatal Python error: (pygame parachute) Floating Point Exception
Python runtime state: initialized

Thread 0x00007f88392f9740 (most recent call first):
  File "/usr/lib/python3.8/site-packages/tensorflow_core/python/eager/execute.py"
``` 
Any Idea about that? 

Anyway, regarding point `1` I think that is possible to decouple model definition from creation. In fact, I just realised that a keras model is build (i.e. weights allocated and initialised) only when a model instance is created, i.e. `Model(inputs, outputs)`. What happens is the following (I assume using keras functional API):
```python
# Model definition:
inputs = Input(shape=...)
x = Dense(...)(inputs)
# .. some more layers
outputs = Dense(...)(x)

# Model creation:
model = Model(inputs, outputs)
```
I think that the above doesn't hold for pretrained architectures (i.e. the ones from `keras.applications`). So, as you said before about point `2`, a `KerasBlock` or `KerasModel` class (I mean something that wraps an already built model) should help in these cases as well.

To sum up, `KerasNetwork` should only handle keras model definition while `KerasBlock/Model` should deal with predefined and/or pretrained models, lazily loading them when tensorforce is building the entire model.

**Comment by AlexKuhnle** [comment]
I have no idea why pygame would complain about eager execution, or why it would even get to know about it? Does that mean TensorFlow is referenced somewhere inside pygame?

Agree re KerasNetwork/Block/Model. Are you planning to experiment, or already experimenting, with it? Otherwise I can potentially look a bit into it over the weekend.

**Comment by Luca96** [comment]
Unfortunately, I'll be busy for 2/3 weeks from now, so I probably won't have the time for any experiment. I'm sorry. Anyway, I can still provide you some support on keras, if needed.

**Comment by AlexKuhnle** [comment]
Closing this since this is quite old by now and loading/pretraining have changed/improved. Re-open with updated issue title and details if required.
